applications/Chat/evaluate/gpt_evaluate.py

```diff
@@ -0,0 +1,519 @@
+
+    Returns:
+        All the evaluations of the given answers.
+    """
+
+    prompt_per_category = {prompt["category"]: prompt for prompt in prompts}
+
+    data_per_category = {}
+    for answer in answers:
+        category = answer["category"]
+
+        if answer["category"] in data_per_category.keys():
+            data_per_category[category].append(answer)
+        else:
+            data_per_category[category] = [answer]
+
+    categories_str = ", ".join(x for x in list(data_per_category.keys()))
+    print(f"The evaluated categories are {categories_str}.")
+
+    if not os.path.exists(save_path):
+        os.makedirs(save_path)
+
+    all_evaluations = []
+
+    for category, data in data_per_category.items():
+        if prompt_per_category.get(category) == None:
+            print(f"No metrics for category {category}! Use category general now.")
+            prompt_per_category[category] = prompt_per_category["general"]
+
+        if os.path.exists(os.path.join(save_path, model_name, f"{category}_evaluation_results.json")):
+            print(
+                f"The evaluation file for category {category} already exists. You are now re-evaluating category {category}!"
+            )
+
+        print(f"The number of instances of category {category}'s is {len(data)}.")
+
+        evaluations = []
+
+        metrics_str = ", ".join(x for x in list(prompt_per_category[category]["metrics"].keys()))
+        print(f"Category {category}'s metrics are {metrics_str}.")
+
+        with concurrent.futures.ThreadPoolExecutor(max_workers=4) as executor:
+            futures = []
+            for inst in data:
+                future = executor.submit(get_gpt35_evaluation, prompt_per_category[category], inst, 1)
+                futures.append(future)
+
+            for future in tqdm.tqdm(
+                    concurrent.futures.as_completed(futures),
+                    desc=f"{category}: ",
+                    total=len(futures),
+            ):
+                evaluations.append(future.result())
+
+        evaluations.sort(key=lambda x: x["id"])
+
+        jdump(
+            evaluations,
+            os.path.join(save_path, model_name, f"{category}_evaluation_results.json"),
+        )
+        print(f"{category} done.")
+
+        all_evaluations.extend(evaluations)
+
+    jdump(
+        all_evaluations,
+        os.path.join(save_path, f"{model_name}_evaluation_results.json"),
+    )
+
+    return all_evaluations
+
+
+def calculate_scores_form_logprobs(logprobs: Dict[str, Any]) -> float:
+    """
+    Calculate score from log probabilities returned by text-davinci-003.
+    Only openai.Completion can return logprobs.
+
+    Args:
+        logprobs: logprobs returned by openai.Completion.
+
+    Returns:
+        Score of one answer.
+    """
+
+    prob = np.zeros(5)
+
+    for key, value in logprobs.items():
+        if "bytes" in key:
+            continue
+        results = re.findall(r"\d", key)
```

Comment:
This function is really hard to understand. Please add some comments or links to illustrate this function.

Reply:
Ok, i will add more comments.